pkg/operator/ceph/cluster/cluster.go

```diff
@@ -52,6 +58,68 @@ func newCluster(c *cephv1beta1.Cluster, context *clusterd.Context) *cluster {
 		ownerRef: ClusterOwnerRef(c.Namespace, string(c.UID))}
 }
 
+func (c *cluster) detectCephMajorVersion() error {
+	// get the major ceph version by running "ceph --version" in the ceph image
+	job := &batch.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      detectVersionName,
+			Namespace: c.Namespace,
+		},
+		Spec: batch.JobSpec{
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"job": detectVersionName,
+					},
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						{
+							Command: []string{"ceph"},
+							Args:    []string{"--version"},
+							Name:    "version",
+							Image:   c.Spec.CephVersion.Image,
+						},
+					},
+					RestartPolicy: v1.RestartPolicyOnFailure,
+				},
+			},
+		},
+	}
+	k8sutil.SetOwnerRef(c.context.Clientset, c.Namespace, &job.ObjectMeta, &c.ownerRef)
+
+	// run the job to detect the version
+	if err := k8sutil.RunReplaceableJob(c.context.Clientset, job); err != nil {
+		return fmt.Errorf("failed to start version job. %+v", err)
+	}
+
+	if err := k8sutil.WaitForJobCompletion(c.context.Clientset, job); err != nil {
+		return fmt.Errorf("failed to complete version job. %+v", err)
+	}
+
+	log, err := k8sutil.GetPodLog(c.context.Clientset, c.Namespace, "job="+detectVersionName)
+	if err != nil {
+		return fmt.Errorf("failed to get version job log to detect version. %+v", err)
+	}
+
+	version, err := extractCephVersion(log)
+	if err != nil {
+		return fmt.Errorf("failed to extract ceph version. %+v", err)
+	}
+
+	logger.Infof("detected ceph version %s for image %s", version, c.Spec.CephVersion.Image)
+	if c.Spec.CephVersion.Name != "" && c.Spec.CephVersion.Name != version {
+		// if the cephVersion.name is already set, override the detected version
+		logger.Warningf("overriding ceph version to %s specified in the CRD", c.Spec.CephVersion.Name)
+	} else {
+		c.Spec.CephVersion.Name = version
+	}
+
+	// delete the job since we're done with it
+	k8sutil.DeleteBatchJob(c.context.Clientset, c.Namespace, job.Name, false)
```

Comment:
Is there any risk of resource leak if one of the `return fmt.Errorf(...)` lines is hit and this `DeleteBatchJob` doesn't fire? Or is this just a formality since the job is already completed?

Reply:
Correct, it's just a formality to delete this job. If there were a failure and the job wasn't deleted, it would be deleted the next time the operator restarts or re-orchestrates the cluster.